### ISSUE TITLE
[Bitnami/Thanos] Question/Issue about Prometheus HA

### ISSUE THREAD

**Issue by dbrennand** [bitnami/charts#3624]
**Which chart**:
Bitnami/Thanos - Version: thanos-2.0.0 - 0.14.0

**Describe the bug**

I am experiencing behaviour where only results from one prometheus replica is showing up when using the Thanos querier.

I have 3 replicas deployed using version `prometheus-operator-0.31.0      0.41.0` of the Prometheus-operator (now named kube-prometheus).

Why don't the other two prometheus-replicas and their scraped results show up in the Thanos querier? Is there something obvious I have missed here?

**Expected behavior**
Scraped results for all 3 prometheus_replicas show.

**Version of Helm and Kubernetes**:

- Output of `helm version`:

```
version.BuildInfo{Version:"v3.3.0", GitCommit:"8a4aeec08d67a7b84472007529e8097ec3742105", GitTreeState:"dirty", GoVersion:"go1.14.7"}
```

- Output of `kubectl version`:

```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.8", GitCommit:"9f2892aab98fe339f3bd70e3c470144299398ace", GitTreeState:"clean", BuildDate:"2020-08-13T16:12:48Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.9", GitCommit:"d2ea02d7bcf37724adabbdb9d37ede0cf2c0d627", GitTreeState:"clean", BuildDate:"2020-04-24T22:59:19Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

**Additional context**
I do have `--query.replica-label "prometheus_replica" \` set, is this the mistake I have made here? When I don't specify deduplication I still only see results for one Prometheus replica.

**Comment by dbrennand** [comment]
This questions stems from an issue I experienced where the first replica of Prometheus began crash looping. 

With the Thanos querier endpoint configured in Grafana during the time of this first replica crashing, no metrics were available.

**Comment by dbrennand** [comment]
![image](https://user-images.githubusercontent.com/52419383/92404399-65153d80-f12b-11ea-93a0-c5af3d14b12f.png)

Is it something to do with only one thanos-sidecar container showing in the announced LabelSets?

**Comment by carrodher** [comment]
Hi,

The parameter to set the replica label is `replicaLabel` instead of `replica-label`, not sure if this is going to be the source of the issue. See https://github.com/bitnami/charts/tree/master/bitnami/thanos#thanos-querier-parameters:
| Parameter                                       | Description                                                                                            | Default                                                 |
|-------------------------------------------------|--------------------------------------------------------------------------------------------------------|---------------------------------------------------------|
| `querier.replicaLabel`                          | Replica indicator(s) along which data is deduplicated                                                  | `replica`                                               |

Apart from that, in order to try to reproduce the issue on our side, can you please provide the parameters (with `--set`) or the custom _values.yaml_ you are using to deploy both charts (Prometheus and Thanos)?

**Comment by dbrennand** [comment]
Hi @carrodher 

Here is the information you requested.

## Chart versions deployed
```
NAME                    NAMESPACE       REVISION        UPDATED                                 STATUS          CHART                           APP VERSION
prometheus-operator     monitoring      1               2020-08-18 09:55:52.2567728 +0100 BST   deployed        prometheus-operator-0.31.0      0.41.0
thanos                  monitoring      1               2020-08-18 10:54:40.1726101 +0100 BST   deployed        thanos-2.0.0                    0.14.0
```

### Thanos values.yaml
```yaml
objstoreConfig: |-
  type: AZURE
  config:
    storage_account:
    storage_account_key:
    container:
    max_retries: 0
querier:
  replicaLabel: prometheus_replica
  dnsDiscovery:
    sidecarsService: prometheus-operator-prometheus-thanos
    sidecarsNamespace: monitoring
bucketweb:
  enabled: true
compactor:
  enabled: true
storegateway:
  enabled: true
ruler:
  enabled: false
metrics:
  enabled: true
  serviceMonitor:
    enabled: true
    namespace: monitoring
```

### Prometheus operator values.yaml
```yaml
operator:
  logLevel: info
prometheus:
  retention: 30d
  logLevel: info
  replicaCount: 3
  scrapeInterval: 1m
  evaluationInterval: 1m
  thanos:
    create: true
    objectStorageConfig:
      secretName: thanos-objstore-config
      secretKey: thanos.yaml
  additionalScrapeConfigsExternal:
    enabled: true
    name: prometheus-operator-prometheus-scrape-config
    key: additional-scrape-configs.yaml
  # Required by Thanos
  disableCompaction: true
  persistence:
    enabled: true
    storageClass: default
    size: 100Gi
alertmanager:
  logLevel: info
  externalConfig: true
  replicaCount: 3
  persistence:
    enabled: true
    storageClass: default
  configMaps:
    - alertmanager-email-template

```

**Comment by carrodher** [comment]
Hi, I am trying to reproduce the issue but as there are many custom configurations I am not able to reproduce the same scenario. Instead of that, I was following [those steps](https://github.com/bitnami/charts/tree/master/bitnami/thanos#integrate-thanos-with-prometheus-and-alertmanager) to deploy a similar scenario, in that case (changing the `replicaCount`) I am not able to see one item per replica in the UI

**Comment by dbrennand** [comment]
> 
> 
> Hi, I am trying to reproduce the issue but as there are many custom configurations I am not able to reproduce the same scenario. Instead of that, I was following [those steps](https://github.com/bitnami/charts/tree/master/bitnami/thanos#integrate-thanos-with-prometheus-and-alertmanager) to deploy a similar scenario, in that case (changing the `replicaCount`) I am not able to see one item per replica in the UI

Hi @carrodher 

Thanks for trying to replicate my setup. I understand that I have some custom config which you cannot replicate.

So to clarify, you were able to replicate the same behaviour in my setup where in the Thanos querier UI only shows LabelSets from one sidecar container?

**Comment by carrodher** [comment]
Apart from the decoupling, the data from the different replicas should appear. I just created an internal task to check this issue. We'll update the thread when we have more information.

**Comment by dbrennand** [comment]
Thanks @carrodher really appreciate the all the help.

**Comment by fsschmitt** [comment]
@carrodher and @dbrennand I was having the similar "issue" initially but I've managed to understand the issue and "solve" it.

Let me explain the reasoning behind it first, so that it would help everyone else also in the future.

Within your `prometheus` configuration, where you are creating the Thanos sidecar (per replica) you are also creating a service specifically for the thanos sidecar (which will be used as a source in the Thanos Querier). Which can be seen the default values [here](https://github.com/bitnami/charts/blob/master/bitnami/kube-prometheus/values.yaml#L734):

```yaml
...
  thanos:
...
    ## Thanos Sidecar Service
    ##
    service:
      ## Kubernetes service type and port number
      ##
      type: ClusterIP
      port: 10901
      # clusterIP: None
...
```

This will generate a service which points to all pods with the following selector:
```yaml
  selector:
    app: prometheus
    prometheus: {{ template "kube-prometheus.prometheus.fullname" . }}
```

Here's where core concepts from K8s about services come in handy (we are all continuously learning). A service, normally is a load-balancer across all pods that match such selector, although there's also a concept named [Headless Service](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services) which makes a service instead of act as a load-balancer, it returns all IP to all pods that match the selector:
> For headless Services that define selectors, the endpoints controller creates Endpoints records in the API, and modifies the DNS configuration to return records (addresses) that point directly to the Pods backing the Service.

Meaning that it will return all sidecars from all prometheus replicas, therefore all we need to do is turn this service into a Headless Service by setting `clusterIP: None`.

### TL;DR

Turn the thanos sidecar service into a headless service and you'll see all sidecar replicas:
```yaml
prometheus:
...
  thanos:
    create: true
    objectStorageConfig:
      secretName: thanos-objstore-config
      secretKey: thanos.yaml
    service:
      type: ClusterIP
      port: 10901
      clusterIP: None
```

**Comment by dbrennand** [comment]
Hi @fsschmitt 

Thank you for the detailed explanation :-) I will try this in my setup and report back.

If you wouldn't mind, could you please provide the YAML representation of the service?

Many Thanks.

**Comment by fsschmitt** [comment]
`spec.clusterIP` is immutable in Kubernetes, so you might have to remove the service and re-create it again.

Can you delete the service (`kubectl delete svc <prometheus_thanos_sidecar_service_name>`, then do a helm upgrade for the prometheus release while changing the values to the following (based on the `values.yaml` you've shared previously):
```yaml
operator:
  logLevel: info
prometheus:
  retention: 30d
  logLevel: info
  replicaCount: 3
  scrapeInterval: 1m
  evaluationInterval: 1m
  thanos:
    create: true
    objectStorageConfig:
      secretName: thanos-objstore-config
      secretKey: thanos.yaml
    service:
      type: ClusterIP
      port: 10901
      clusterIP: None
  additionalScrapeConfigsExternal:
    enabled: true
    name: prometheus-operator-prometheus-scrape-config
    key: additional-scrape-configs.yaml
  # Required by Thanos
  disableCompaction: true
  persistence:
    enabled: true
    storageClass: default
    size: 100Gi
alertmanager:
  logLevel: info
  externalConfig: true
  replicaCount: 3
  persistence:
    enabled: true
    storageClass: default
  configMaps:
    - alertmanager-email-template
```

Please note that the only difference is adding on the thanos section the:
```yaml
    service:
      type: ClusterIP
      port: 10901
      clusterIP: None
```

**Comment by carrodher** [comment]
Thank you very much @fsschmitt for your detailed explanation about how to fix the issue. I'm glad that all of you were able to solve the issue with the above workaround.

cc @juan131 as you was looking/interesting on this issue

**Comment by juan131** [comment]
Thanks so much for the detailed explanation @fsschmitt !! I had this in my to-do list but I hadn't found the time to look into it yet. 

IMHO, I think we should use a headless service for Thanos sidecar's service by default. That's why I created the PR above.

**Comment by fsschmitt** [comment]
@juan131, I wouldn't necessarily say that having `the` service being deployed as a Headless service would be the most K8s friendly approach as this is a Service Discovery feature. 

Have you thought about implementing the same way as plenty of other charts (e.g. Elasticsearch, Thanos, etc.) have, which deploys both a normal K8s service (load balancing the requests between the pods) and a Headless service (for service discovery usage). 

https://github.com/bitnami/charts/blob/master/bitnami/thanos/templates/storegateway/service.yaml
https://github.com/bitnami/charts/blob/master/bitnami/thanos/templates/storegateway/service-headless.yaml

I'm sure there are non-thanos querier use cases where a service discovery is not required. 

Nonetheless, I would avoid changing the defaults on a bitnami helm chart as it could break functionality on people who are using the default values (wrongly) and using the service as a LB instead for service discovery.

Would rather add to the `Thanos Querier` section on the `Thanos helm chart` a note on the documentation explaining what would expect from a Thanos Sidecar service if the goal would be to use service discovery instead of using direct IPs or direct kube-dns.

**Comment by juan131** [comment]
Hi @fsschmitt 

I agree with being careful to avoid introducing backwards incompatibilities. I'll evaluate the consequences of upgrading `kube-prometheus` if the ClusterIP changes.

Regarding the convenience of using service discovery for this specific feature, I can't think of any use case since this Thanos sidecar container (and corresponding service) is directly associated with the use of Thanos. I mean, it has no other purpose. 

Thanos Querier needs to query data from every sidecar container and, therefore, that's why I think **it should be a headless svc by default**.

**Comment by fsschmitt** [comment]
@juan131 I'm not disagreeing, just raising awareness for any other use-case that also myself couldn't come up with 🙂

**Comment by juan131** [comment]
@fsschmitt as you predicted... This change will turn out to fail during upgrades due to immutable fields:

```console
Error: UPGRADE FAILED: cannot patch "prometheus-operator-prometheus-thanos" with kind Service: Service "prometheus-operator-prometheus-thanos" is invalid: spec.clusterIP: Invalid value: "None": field is immutable
```

But it's easy to workaround it by removing the service and let helm recreate it during the upgrade:

```
$ helm install prometheus-operator \
    --set prometheus.thanos.create=true \
    --namespace monitoring \
    bitnami/prometheus-operator
$ kubectl delete svc prometheus-operator-prometheus-thanos -n monitoring
$ helm upgrade --install prometheus-operator \
    --set prometheus.thanos.service.clusterIP=None \
    --set prometheus.thanos.create=true \
    --namespace monitoring \
    bitnami/prometheus-operator
```

I'd include these instructions in the README.md, so users facing issues can easily upgrade their charts.
